lib/kafka/async_producer.rb

```diff
@@ -288,6 +275,29 @@ def do_loop
         retry
       end
 
+      def shutdown
+        # Deliver any pending messages first.
+        @producer.deliver_messages
```

Comment:
Is it worth logging how many messages are to be delivered as part of shutdown?

Reply:
We are already collecting that info normally, so the question is are we learning anything extra by knowing when the batch is part of shutdown? Maybe. And I don't think that change would be tacked onto this PR.

Reply:
Also should the queue be closed as part of shutdown regardless of whether there was an error caught or not?

Reply:
Maybe we shouldn't call `#close` at all. The original code did not, it was an addition by @BWStearns and I don't think it was strictly necessary. @BWStearns was there a specific reason for closing the queue?

Reply:
Looking at the old pr now, just saw this comment.

Reply:
I guess close isn't strictly necessary since nothing should be adding to it during that time. I think I did it to ensure that `queue.size.times` would always match the size of the queue (but again this shouldn't be necessary). There is one behavior though that might have made me do that. A call to pop on an open but empty queue will just block forever, whereas a closed queue will return `nil` immediately. Closing it ensures that you have a way of knowing that the queue is empty.

Reply:
Hmm, in that case, I'll close it at shutdown. This will trigger a `ClosedQueueError` though, which we are currently not handling.

Reply:
It seems like that `ClosedQueueError` should be handled here. imo a client would expect `shutdown` not to throw any errors because it's ending the lifecycle of the producer and nothing can be done about it anyway.

Reply:
The `ClosedQueueError` would happen if another client tried to produce after the shutdown. There could also potentially be a client producing on another thread.

Perhaps we should wrap the error here with a `Kafka::Error` so that it seems like it came from Kafka and existing `rescue` logic might handle it. Thoughts?

Reply:
That seems like a good idea to me because a caller can do a catchall on the more generic `Kafka::Error`. If this becomes a common but harmless occurrence it would be nice to be able to exclude it from Sentry somehow.

Reply:
What about calling pop with nonblock and catch the ThreadError?

```
irb(main):002:0> q.pop(true)
(irb):2:in `pop': queue empty (ThreadError)
```

Reply:
It avoids closing the queue and gives an identifiable empty state (the ThreadError being thrown)

Reply:
`ThreadError`????!?!? I'd want to take a look at the source for `Queue` to see why it does that.

Reply:
It is indeed not the error one would imagine (especially given that they bothered implementing specific errors for other queue issues).